### ISSUE TITLE
Adopt TypeScript for serverside code

### ISSUE THREAD

**Issue by Turbo87** [cardstack/cardstack#472]
While working on #466 yesterday I discovered that there are multiple `Writer` implementations that all seem to implement a common interface. Unfortunately that interface doesn't seem to be specified or documented anywhere.

I believe that incrementally adopting TypeScript for at least the serverside code would be quite beneficial to resolving these kinds of issues. I would propose to evaluate what infrastructure would be needed to support TypeScript on the serverside (especially in combination with lerna, etc.), and then incrementally adopt it.

**Comment by ef4** [comment]
Yes, we have discussed in the past and everybody is on board. Somebody just needs to prioritize working on the build infrastructure.

tsconfig makes it somewhat annoying to build a whole monorepo at once while keeping the TS sources and the JS output in their own per-package subdirectories. 

For comparisons, the glimmer monorepo does a custom build pipeline with broccoli. And in the embroider monorepo I just accepted that TS is going to put the build output side-by-side with the real sources, which is not great (you see a ton of .js, .d.ts, and .map files in your editor sidebar, though gitignoring them helps keep them out of most searches, etc).
